Sources/BSON/BSONDecimal128.swift

```diff
@@ -0,0 +1,307 @@
+import Foundation
+import NIO
+
+public struct UInt128: Equatable, Hashable {
+    /// The high order 64 bits
+    public var hi: UInt64
+    /// The low order 64 bits
+    public var lo: UInt64
+
+    public static func multiply(_ left: UInt64, by right: UInt64) -> UInt128 {
+        let (product, didOverflow) = left.multipliedReportingOverflow(by: right)
+        if !didOverflow {
+            return UInt128(hi: 0, lo: product)
+        }
+
+        let leftHi = UInt64(UInt32(left >> 32))
+        let leftLo = UInt64(UInt32(left) & UInt32.max)
+
+        let rightHi = UInt64(UInt32(right >> 32))
+        let rightLo = UInt64(UInt32(right) & UInt32.max)
+
+        var productHi = leftHi * rightHi
+        var productMid0 = leftHi * rightLo
+        let productMid1 = leftLo * rightHi
+        var productLo = leftLo * rightLo
+
+        productHi += productMid0 >> 32
+        productMid0 += productMid1 + (productLo >> 32)
+
+        productHi += productMid0 >> 32
+        productLo += productMid0 << 32
+
+        return UInt128(hi: productHi, lo: productLo)
+    }
+}
+
+public struct BSONDecimal128: BSONValue, Equatable, Hashable, CustomStringConvertible {
+    private static let DIGITS_RE = #"(?:\d+)"#
+    private static let INDICATOR_RE = #"(?:e|E)"#
+    private static let SIGN_RE = #"[+-]"#
+    private static let INFINITY_RE = #"Infinity|Inf|infinity|inf"#
+    private static let DECIMAL_PART_RE = "\(DIGITS_RE)\\.\(DIGITS_RE)?|\\.?\(DIGITS_RE)"
+    private static let NAN_RE = #"NaN"#
+    // swiftlint:disable line_length
+    private static let EXPONENT_PART_RE = "\(INDICATOR_RE)(\(SIGN_RE))(\(DIGITS_RE))"
+    private static let NUMERIC_VALUE_RE = "(\(SIGN_RE))?(?:(\(DECIMAL_PART_RE))(?:\(EXPONENT_PART_RE))?|(\(INFINITY_RE)))"
+    // swiftlint:enable line_length
+    public static let DECIMAL128_RE = "\(NUMERIC_VALUE_RE)|(\(NAN_RE))"
+
+    private static let PRECISION_DIGITS = 34
+    private static let EXPONENT_MAX = 6144
+    private static let EXPONENT_BIAS = EXPONENT_MAX + PRECISION_DIGITS - 2
+    private static let EXPONENT_MIN = 1 - EXPONENT_MAX
+
+    private static let NEG_INFINITY = UInt128(hi: 0xF800_0000_0000_0000, lo: 0)
+    private static let INFINITY = UInt128(hi: 0x7800_0000_0000_0000, lo: 0)
+    private static let NAN = UInt128(hi: 0x7C00_0000_0000_0000, lo: 0)
+
+    private enum REGroups: Int, CaseIterable {
+        case sign = 1
+        case decimalPart = 2
+        case exponentSign = 3
+        case exponentPart = 4
+        case infinity = 5
+        case nan = 6
+    }
+
+    public enum BSONDecimal128Errors: Error {
```

Comment:
I think switching the errors thrown to a new type is a breaking change for the driver, so we probably can't use this approach. I'd suggest defining some constants corresponding to each of these cases that are `InvalidArgumentError`s (or whatever the appropriate type would be given the context you're throwing them in)